packages/sphinx-external-toc/src/toc.ts

```diff
@@ -0,0 +1,168 @@
+import yaml from 'js-yaml';
+import type { TOC, ArticleTOC, BookTOC, BasicTOC } from './types.js';
+import schema from './schema.json';
+import Ajv from 'ajv';
+
+// See https://executablebooks.org/en/latest/blog/2021-06-18-update-toc/
+// Implementation transpiled from https://github.com/executablebooks/sphinx-external-toc/blob/v1.0.1/sphinx_external_toc/tools.py#L277
+function upgradeOldJupyterBookTOC(oldTOC: any[]) {
+  const tocUpdated = oldTOC[0] as Record<string, unknown>;
+
+  let firstItems: Record<string, unknown>[] = [];
+  let topItemsKey = 'sections';
+
+  if ('sections' in tocUpdated && 'chapters' in tocUpdated) {
+    throw new Error("First list item contains both 'chapters' and 'sections' keys");
+  }
+
+  // Identify whether we have "sections" or "chapters" in our first key
+  // And pull them into "firstItems" if so
+  for (const key of ['sections', 'chapters']) {
+    if (key in tocUpdated) {
+      topItemsKey = key;
+      const items = tocUpdated[key] as Record<string, unknown>[];
+      delete tocUpdated[key];
+
+      if (!Array.isArray(items)) {
+        throw new Error();
```

Comment:
```suggestion
        throw new Error(`'${key}' in toc must be an array`);
```

Reply:
Fab, this one was left as a TODO :facepalm: